README.md
## A plugin for [Dragory's ModMail](https://github.com/dragory/modmailbot) that allows users to open threads with reactions  
**Currently on Version 1.3**  
A full [changelog can be found here](https://github.com/MMPlugins/ReactionThreads/blob/main/CHANGELOG.md).  
Plugin written and maintained by [DarkView](https://github.com/DarkView) (Dark#1010 on Discord)  
  
Table of Contents:
- [Setup](#setup)
- [Usage](#usage)
  - [General](#general)
  - [Restrictions](#restrictions)
  - [Bugs](#bugs)
  - [Feature Requests](#feature-requests)
- [Commands](#commands)
    - [Adding/Registering a reaction](#addingregistering-a-reaction)
    - [Removing/De-Registering a reaction](#removingde-registering-a-reaction)  
    - [Adding/Removing/Displaying custom responses](#addingremovingdisplaying-custom-responses)  
    - [Listing all reactions](#listing-all-reactions)  
    - [Refreshing all reactions](#refreshing-all-reactions)  
    
## Setup
Make sure you are running at least v3.3.0 of Modmail.
in your config.ini file, create a new line and add  
```
plugins[] = npm:MMPlugins/ReactionThreads
reactionThreads-ownerId = <yourUserId or adminRoleId>
extraIntents[] = guildMessageReactions
```
to your config and replace `<yourUserId>` with your user ID. Only the person with this ID is allowed to add/remove reactions.  
Grant the ModMail bot "Manage Message" permissions in any channel you want it to use reactions for, as it removes the reaction the user adds automatically.  
You need to restart the bot in order for the plugin to be loaded!

## Usage
### General
In order for the bot to create threads through user reactions, you need to set up reactions first. See the Commands section below for information on how to do so.  
Registered reactions are persistent between restarts, as long as you do not delete or move the `ReactionThreadsData.json` file created by the plugin.
### Restrictions
The plugin is rather lenient regarding allowed reactions. As long as the bot is able to use the emoji, you are able to register it as a reaction.
  
It is not possible to create more than one thread per user, regardless of how it is created. If the user already has an active thread, the reaction will simply be removed again without anything happening.  
  
In the case you run mutliple bot instances from the same folder, you can set the `reactionThreads-suffix` config option in your config.ini. This will append the given text to the `ReactionThreadsData.json` file so the different instances can use different reaction lists.
### Bugs
There are currently no known bugs.  
If you have found a bug, please report it at the [issues page for the plugin](https://github.com/MMPlugins/ReactionThreads/issues)!  
You can also find the plugin author (Dark#1010) on the [official support discord](https://discord.gg/vRuhG9R) in case you have any specific questions.
### Feature Requests
If you want to request or suggest a feature, open an issue on the [plugins issue page](https://github.com/MMPlugins/ReactionThreads/issues)!
In case the feature you want to request is outside of the scope of this plugin (anything not to do with reactions creating threads) please use the `#plugin-requests` channel on the [official support discord](https://discord.gg/vRuhG9R).
## Commands

Parameters in <> are required, parameters in [] are optional.  
These commands only work on the inbox server, just like regular ModMail commands.  
### Adding/Registering a reaction
Signature 1: `!rtAdd <ChannelID> <MessageID> <Emoji> [CategoryID]`  
Signature 2: `!rtAdd <ChannelID> <MessageID> <Emoji> <CategoryID> [PingRoleID]`  
This will register a new reaction and immediately activate it, making any reactions to it create a new thread.
- `ChannelID` has to be the ID of the channel the message the reaction should be added to is in.
- `MessageID` is the ID of the message in that channel.
- `Emoji` is just the emoji directly from the emoji picker without any changes.
- `CategoryID` should only be used if you want reactions to that message to be placed in a specific category.
- `PingRoleID` can only be used in conjunction with `CategoryID`. It allows you to set a role to be pinged when a thread is created with that specific reaction.

### Removing/De-Registering a reaction
Signature: `!rtRemove <ChannelID> <MessageID> <Emoji>`  
This will de-register an already existing reaction. All threads opened via the reaction remain open.  
Takes the same arguments as `!rtAdd` but without the optional parameters, as that is not needed for reaction matching (cannot have the same emoji twice).  

### Adding/Removing/Displaying custom responses
Signature: `!rtResponse <ChannelID> <MessageID> <Emoji> [Response]`  
Takes the same parameters as `!rtRemove` and also an optional `Response` parameter.  
If the optional parameter is *not* passed, the bot will display the currently set custom response.  
If the optional parameter *is* passed, the bot will set the reaction to use this new response.
- `Response` can by any string, including emojis. To use emoji, the bot must be on the server the emoji is from. Besides this, the only thing that limits a response length is a maximum of 1500 characters.  
If you want to completely remove a response, pass one of the following without anything else: `none`, `nothing`, `empty`, `null`, `-`.  

This command and its functionality was made by [YetAnotherConnor](https://github.com/YetAnotherConnor).
  
### Listing all reactions
Signature: `!rtList [Any ID]`  
This will return a list of all current reactions that match the ID filter if one is given.  
- `Any ID` can be any of the IDs available for reactions: `Channel ID`, `Message ID` and `Category ID`. If this parameter is passed, only reactions that are on that message or in that channel etc. get listed.  

### Refreshing all reactions
Signature: `!rtRefresh`  
This will refresh all reactions this plugin manages.  
This happens every 30 minutes automatically, and this command advances the next refresh to happen immediately.


Table of Contents: 
- [Setup](#setup)
- [Usage](#usage)
  - [General](#general)
  - [Restrictions](#restrictions)
  - [Bugs](#bugs)
  - [Feature Requests](#feature-requests)
- [Commands](#commands)
    - [Adding/Registering a reaction](#addingregistering-a-reaction)
    - [Removing/De-Registering a reaction](#removingde-registering-a-reaction)  
    - [Adding/Removing/Displaying custom responses](#addingremovingdisplaying-custom-responses)  
    - [Listing all reactions](#listing-all-reactions)  
    - [Refreshing all reactions](#refreshing-all-reactions)
  
Plugin written and maintained by [DarkView](https://github.com/DarkView) (Dark#1010 on Discord)  


CHANGELOG.md
## Version 1.3
- New command: `rtRefresh`  
Allows any owner to instantly refresh all reactions  
  
The bot now removes and re-applies all reactions it manages automatically every 30 minutes and on restart.  
In case you want to immediately refresh, you can use the `rtRefresh` command which advances the next check so it happens right away.
  
Please read the [command documentation](https://github.com/MMPlugins/ReactionThreads#commands) to learn how to use this new command!

## Version 1.2
- New command: `rtList`  
Allows any owner to list all reactions the bot currently has registered
- New command: `rtResponse` by [YetAnotherConnor](https://github.com/YetAnotherConnor)  
Allows you to set a custom response for specific reactions which will be sent instead of the regular response message.  
  
Please read the [command documentation](https://github.com/MMPlugins/ReactionThreads#commands) to learn how to use these new commands!

## Version 1.1
- `reactionThreads-ownerId` can now be a role id.  
If you want everyone with a specific role to be able to add reactions, set the value to the given role's ID.
- Plugin now checks if there was an update at startup (npm downloads new plugin version on startup)  
If a new version is installed, tells user to check the changelog here.


reactionThreads.js
module.exports = function ({ bot, config, commands, knex, threads }) {
  const fs = require("fs");
  const pluginVersion = "1.3";
  const changelogUrl = "=> https://daark.de/RTCL <=";
  let reactions = [];
  const emptyResponse = ["none", "nothing", "empty", "null", "-"];
  let refreshTimeout = null;

  // Check if ownerId is specified in the config, warn otherwise
  const ownerId = config["reactionThreads-ownerId"];
  if (typeof ownerId === "undefined") {
    console.info(
      '[ReactionThreads] No ownerId specified in config via "reactionThreads-ownerId", everyone (with inboxServerPermission) will be able to add new reactions!',
    );
  }
  // Load the suffix for the json file, if one exists (used for multiple bot instances running from the same folder)
  const jsonSuffix = config["reactionThreads-suffix"] ? config["reactionThreads-suffix"] : "";

  // Warn the user not to delete the file in case it doesn't exist (basically a first-use check)
  if (!fs.existsSync(`./ReactionThreadsData${jsonSuffix}.json`)) {
    console.info(
      `[ReactionThreads] A ReactionThreadsData${jsonSuffix}.json file will be created when using this plugin. Please do not modify or delete this file or reactions you set up will cease to function.`,
    );
  } else {
    // Load registered reactions if the file exists
    const data = fs.readFileSync(`./ReactionThreadsData${jsonSuffix}.json`);
    reactions = JSON.parse(data);
    console.info(`[ReactionThreads] Successfully loaded ${reactions.length - 1} reaction(s)`);
  }

  /**
   * Stores all registered reactions into the data file for persistence
   */
  const saveReactions = function () {
    fs.writeFileSync(`./ReactionThreadsData${jsonSuffix}.json`, JSON.stringify(reactions));
  };

  /**
   * Checks whether userId is blocked
   * @param {String} userId
   * @returns {Promise<Boolean>}
   */
  async function isBlocked(userId) {
    const row = await knex("blocked_users")
      .where("user_id", userId)
      .first();
    return !!row;
  }
  
  /**
   * Checks whether or not passed parameters are a valid reaction
   * @param {string} channelId The ID of the channel for which to check
   * @param {string} messageId The ID of the message for which to check
   * @param {string} emoji The stringified emoji for which to check (i.e. <:test:108552944961454080>)
   * @returns full reaction if valid, null if not
   */
  const isValidReaction = function (channelId, messageId, emoji) {
    for (const reaction of reactions) {
      if (reaction.channelId == channelId && reaction.messageId == messageId && reaction.emoji == emoji) {
        return reaction;
      }
    }
    return null;
  };

  /**
   * Checks whether or not passed parameters has response
   * @param {string} channelId The ID of the channel for which to check
   * @param {string} messageId The ID of the message for which to check
   * @param {string} emoji The stringified emoji for which to check (i.e. <:test:108552944961454080>)
   * @returns full reaction if valid, null if not
   */
  const hasResponse = function (channelId, messageId, emoji) {
    const reaction = isValidReaction(channelId, messageId, emoji);
    if (reaction && reaction.response) {
      return reaction;
    }
    return null;
  };

  /**
   * Checks whether or not the user invoking a command is authorized or not
   * @param {*} msg the message to check permissions for
   */
  const isOwner = function (msg) {
    if (typeof ownerId === "undefined") return true;
    return msg.member.id === ownerId ? true : msg.member.roles.includes(ownerId);
  };

  /**
   * Registers a new reaction for use
   * @param {Message} msg The message invoking the command
   * @param {*} args The arguments passed (check registering at bottom)
   */
  const addReactionCmd = async (msg, args) => {
    if (!isOwner(msg)) return;
    if (isValidReaction(args.channelId, args.messageId, args.emoji)) {
      msg.channel.createMessage(`⚠️ Unable to add reaction: That reaction already exists on that message!`);
      return;
    }
    args.categoryId = args.categoryId ? args.categoryId : null;

    try {
      // Replace the trailing > because eris filters out the rest
      await bot.addMessageReaction(args.channelId, args.messageId, args.emoji.replace(">", ""));
    } catch (e) {
      msg.channel.createMessage(
        `⚠️ Unable to add reaction: \`${e}\`\nPlease ensure that the IDs are correct and that the emoji is from one of the servers the bot is on!`,
      );
      return;
    }

    reactions.push({ ...args });
    saveReactions();
    msg.channel.createMessage("Successfully added reaction to message and registered it internally.");
  };

  /**
   * Deregister an existing reaction
   * @param {Message} msg The message invoking the command
   * @param {*} args The arguments passed (check registering at bottom)
   */
  const removeReactionCmd = async (msg, args) => {
    if (!isOwner(msg)) return;
    if (isValidReaction(args.channelId, args.messageId, args.emoji) == null) {
      msg.channel.createMessage(`⚠️ Unable to remove reaction: That reaction doesn't exist on that message!`);
      return;
    }

    try {
      // Replace the trailing > because eris filters out the rest
      await bot.removeMessageReaction(args.channelId, args.messageId, args.emoji.replace(">", ""), bot.user.id);
    } catch (e) {
      msg.channel.createMessage(`⚠️ Unable to remove reaction: \`${e}\``);
    }

    reactions.splice(reactions.indexOf({ ...args }), 1);
    saveReactions();
    msg.channel.createMessage("Successfully removed reaction from message and de-registered it internally.");
  };

  /**
   * Registers or updates a reaction response
   * @param {Message} msg The message invoking the command
   * @param {*} args The arguments passed (check registering at bottom)
   */
  const ReactionRespCmd = async (msg, args) => {
    if (!isOwner(msg)) return;
    reaction = isValidReaction(args.channelId, args.messageId, args.emoji);
    if (reaction == null) {
      msg.channel.createMessage(`⚠️ Unable to add reaction response: That reaction doesn't exist on that message!`);
      return;
    }

    if (args.response) {
      const response = emptyResponse.includes(args.response.trim()) ? null : args.response.trim();
      if (response.length > 1500) {
        msg.channel.createMessage("⚠️ That custom response is too long!");
        return;
      }

      reaction.response = response;
      saveReactions();
      msg.channel.createMessage("Successfully created/updated reaction response and registered it internally.");
    } else {
      msg.channel.createMessage(
        `The current response for that reaction is: \`${reaction.response ? reaction.response : "None"}\``,
      );
    }
  };

  /**
   * Lists all (or specific) reactions
   * @param {*} msg The message which invoked the command
   * @param {*} args The arguments passed to us, i.e. an ID
   */
  const listReactionsCmd = async (msg, args) => {
    if (!isOwner(msg)) return;
    const checkId = args.anyId ? args.anyId.trim() : null;
    let toPost = "Emoji - Channel ID - Message ID - Category ID - First words of response";
    for (const react of reactions) {
      if (react.channelId === "version") continue;
      if (args.anyId && !(react.categoryId === checkId || react.channelId === checkId || react.messageId === checkId))
        continue;

      const reactionText = `\n${react.emoji} - ${react.channelId} - ${react.messageId} - ${
        react.categoryId ? react.categoryId : "None"
      } - ${react.response ? react.response.substring(0, 25) + "..." : "None"}`;

      toPost += reactionText;
    }

    msg.channel.createMessage(toPost);
  };

  /**
   * Handles any reaction added within a guild. If it is a registered reaction, create a thread if none exists
   * @param {Message} message The message that got reacted to
   * @param {Emoji} emoji The emoji used to react
   * @param {Member} reactor The member object of the person reacting
   */
  const onReactionAdd = async (message, emoji, reactor) => {
    if (reactor.user.bot || !reactor.guild) return;
    const stringifiedEmoji = emoji.id ? `<${emoji.animated ? "a" : ""}:${emoji.name}:${emoji.id}>` : emoji.name;
    const reaction = isValidReaction(message.channel.id, message.id, stringifiedEmoji);
    const userHasThread = await threads.findOpenThreadByUserId(reactor.id);
    if (reaction != null && userHasThread == null && !(await isBlocked(reactor.id))) {
      const newThread = await threads.createNewThreadForUser(reactor.user, {
        source: "reaction",
        categoryId: reaction.categoryId,
      });

      const toPing = reaction.pingRoleId != null ? reaction.pingRoleId : null;
      const responseMessage = Array.isArray(config.responseMessage)
        ? config.responseMessage.join("\n")
        : config.responseMessage;
      const postToThreadChannel = config.showResponseMessageInThreadChannel;

      await newThread.postSystemMessage(
        `:gear: **ReactionThreads:** Thread opened because of reaction ${stringifiedEmoji} to https://discord.com/channels/${
          message.channel.guild.id
        }/${message.channel.id}/${message.id}${toPing != null ? " <@&" + toPing + ">" : ""}`,
        { allowedMentions: { roles: [toPing] } },
      );

      newThread
        .sendSystemMessageToUser(reaction.response ? reaction.response : responseMessage, { postToThreadChannel })
        .catch((e) => {
          // Ideally this will be fixed upstream at some point
          newThread.postSystemMessage(
            "⚠️ **ReactionThreads:** Could not open DMs with the user. They may have blocked the bot or set their privacy settings higher.",
          );
        });

      try {
        await bot.removeMessageReaction(message.channel.id, message.id, stringifiedEmoji.replace(">", ""), reactor.id);
      } catch (e) {
        newThread.postSystemMessage(`⚠️ Failed to remove reaction from message: \`${e}\``);
      }
    } else if (reaction != null) {
      // Only remove reaction if the user has an existing thread or user is blocked
      try {
        await bot.removeMessageReaction(message.channel.id, message.id, stringifiedEmoji.replace(">", ""), reactor.id);
      } catch (e) {
        console.error(`[ReactionThreads] Error when trying to remove reaction: ${e}`);
      }
    }
  };

  //#region reaction refresh loop
  const refreshReactions = async (msg = null) => {
    clearTimeout(refreshTimeout);
    if (msg) {
      if (!isOwner(msg)) return;
      msg = await msg.channel.createMessage(`Refreshing all reactions...`);
    }

    for (const react of reactions) {
      if (react.channelId === "version") continue;
      const emoji = react.emoji.replace(">", "");

      await bot.removeMessageReaction(react.channelId, react.messageId, emoji, bot.user.id).catch(() => {});
      await bot.addMessageReaction(react.channelId, react.messageId, emoji).catch(e => {
        console.warn(`[ReactionThreads] Error applying reaction ${react.emoji} to message ${react.channelId}-${react.messageId}: ${e}\nIf you deleted the message, use rtRemove ASAP`);
        if (msg) {
          msg.channel.createMessage(`Error applying reaction ${react.emoji} to message ${react.messageId} in <#${react.channelId}>: ${e}\nIf you deleted the message, use \`rtRemove ${react.channelId} ${react.messageId} ${react.emoji}\``);
        }
      });
    }

    if (msg) {
      await msg.edit(`Done!`);
    }
    refreshTimeout = setTimeout(() => {
      refreshReactions();
    }, 1000 * 60 * 30); //Refresh reactions every 30 minutes 
  }

  refreshReactions();
  //#endregion

  //#region versioncheck
  // Check the plugin version and notify of any updates that happened
  let reactVersion = null;
  for (const reaction of reactions) {
    if (reaction.channelId == "version") {
      reactVersion = reaction;
      break;
    }
  }

  if (reactVersion && reactVersion.messageId != null && reactVersion.messageId != pluginVersion) {
    console.info(
      `[ReactionThreads] Plugin updated to version ${pluginVersion}, please read the changelog at ${changelogUrl} as there may be important or breaking changes!`,
    );
    reactions.splice(reactions.indexOf(reactVersion), 1);
    reactions.push({ channelId: "version", messageId: pluginVersion });
    saveReactions();
  } else if (reactVersion == null) {
    reactions.push({ channelId: "version", messageId: pluginVersion });
    saveReactions();
  }
  //#endregion versioncheck

  //#region registering
  // Register all commands and listeners
  commands.addInboxServerCommand(
    "rtAdd",
    [
      { name: "channelId", type: "string", required: true },
      { name: "messageId", type: "string", required: true },
      { name: "emoji", type: "string", required: true },
      { name: "categoryId", type: "string", required: false },
    ],
    addReactionCmd,
  );

  commands.addInboxServerCommand(
    "rtAdd",
    [
      { name: "channelId", type: "string", required: true },
      { name: "messageId", type: "string", required: true },
      { name: "emoji", type: "string", required: true },
      { name: "categoryId", type: "string", required: true },
      { name: "pingRoleId", type: "string", required: false },
    ],
    addReactionCmd,
  );

  commands.addInboxServerCommand(
    "rtResponse",
    [
      { name: "channelId", type: "string", required: true },
      { name: "messageId", type: "string", required: true },
      { name: "emoji", type: "string", required: true },
      { name: "response", type: "string", required: false, catchAll: true },
    ],
    ReactionRespCmd,
  );

  commands.addInboxServerCommand(
    "rtRemove",
    [
      { name: "channelId", type: "string", required: true },
      { name: "messageId", type: "string", required: true },
      { name: "emoji", type: "string", required: true },
    ],
    removeReactionCmd,
  );

  commands.addInboxServerCommand("rtList", [{ name: "anyId", type: "string", required: false }], listReactionsCmd);

  commands.addInboxServerCommand("rtRefresh", [], refreshReactions);

  bot.on("messageReactionAdd", onReactionAdd);
  //#endregion
};
